### PR TITLE
kvdb/postgres: fix context cancellation

### LINKED ISSUE #6103: Postgres v13.3 or v14.1 backend failing with LND v0.14.1-beta
### Background

I'm able to successfully run a node with bbolt/boltdb. 
When I try to run a node with postgres, after doing a `lncli create` I get an error "Unable to process chain backend block connected notification: context canceled". 

Note it does successfully create the tables before crashing:
- channeldb_kv
- decayedlogdb_kv
- macaroondb_kv
- towerclientdb_kv
- towerserverdb_kv
- walletdb_kv

Can reproduce this with clean environments. Verified Bitcoind node, postgres db is in working order.

### Your environment

* version of `lnd`
`v0.14.1-beta`
* which operating system (`uname -a` on *Nix)
ubuntu 20.04
* version of `btcd`, `bitcoind`, or other backend
bitcoind v22.0
* any other relevant environment details
postgres version 13.3 

### Steps to reproduce
- lnd.conf setup for postgres

```
[db]
db.backend=postgres

[postgres]
db.postgres.dsn=postgres://postgres:PASSWORD@hostname:5432/dbname
db.postgres.timeout=60s
db.postgres.maxconnections=1000
```

- lncli create
- crash with error in logs
### Expected behaviour

No crash

### Actual behaviour

crash `[ERR] LNWL: Unable to process chain backend block connected notification: context canceled` v13.3
or "error loading chain control: unable to create chain control: unable to create wallet: context canceled" v14.1
crash `

Logs when I use bbolt/botdb
```
2021-12-20 03:09:52.444 [INF] LTND: We're not running within systemd
2021-12-20 03:09:52.446 [INF] LTND: Waiting for chain backend to finish sync, start_height=714882
2021-12-20 03:09:53.479 [INF] LNWL: Started rescan from block 0000000000000000000a2daeed9f01f2ca39cd2839026c3f12ceb0f388f815ce (height 714553) for 0 addresses
2021-12-20 03:09:53.502 [INF] LNWL: Catching up block hashes to height 714562, this might take a while
2021-12-20 03:09:53.576 [INF] LNWL: Done catching up block hashes
2021-12-20 03:09:53.576 [INF] LNWL: Rescanned through block 000000000000000000081f53c492703c785f6fff23a92aa7325ff5603682e4f7 (height 714562)
2021-12-20 03:10:15.447 [INF] LNWL: Catching up block hashes to height 714882, this might take a while
2021-12-20 03:10:15.448 [INF] LNWL: Done catching up block hashes
2021-12-20 03:10:15.448 [INF] LNWL: Finished rescan for 0 addresses (synced to block 00000000000000000004a713a1136faaeb84bf34ff2117b6c5c65e7031772d21, height 714882)
2021-12-20 03:10:15.863 [INF] LTND: Chain backend is fully synced (end_height=714882)!
2021-12-20 03:10:15.863 [WRN] HLCK: check: disk space configured with 0 attempts, skipping it
2021-12-20 03:10:15.863 [WRN] HLCK: check: tls configured with 0 attempts, skipping it
2021-12-20 03:10:15.870 [INF] PRNF: PeerNotifier starting
2021-12-20 03:10:15.870 [INF] NTFN: New block epoch subscription
2021-12-20 03:10:15.871 [INF] NTFN: New block epoch subscription
2021-12-20 03:10:15.872 [INF] BRAR: Starting contract observer, watching for breaches.
2021-12-20 03:10:15.874 [INF] NTFN: New block epoch subscription
2021-12-20 03:10:15.874 [INF] DISC: Authenticated Gossiper is starting
2021-12-20 03:10:15.874 [INF] NTFN: New block epoch subscription
2021-12-20 03:10:15.877 [INF] CRTR: FilteredChainView starting
2021-12-20 03:10:15.883 [INF] CRTR: Filtering chain using 0 channels active
2021-12-20 03:10:15.886 [INF] CRTR: Prune tip for Channel Graph: height=714882, hash=00000000000000000004a713a1136faaeb84bf34ff2117b6c5c65e7031772d21
2021-12-20 03:10:15.888 [INF] NTFN: New block epoch subscription
2021-12-20 03:10:15.890 [INF] NTFN: New block epoch subscription
2021-12-20 03:10:15.890 [INF] HSWC: Starting HTLC Switch
2021-12-20 03:10:15.890 [INF] CHFT: Adding 0 channels to event store
2021-12-20 03:10:15.890 [INF] CHBU: Starting chanbackup.SubSwapper
2021-12-20 03:10:15.890 [INF] NTFN: New block epoch subscription
```

Logs when using postgres v13.3
```
2021-12-20 03:22:49.411 [INF] LTND: We're not running within systemd
2021-12-20 03:22:49.413 [INF] LTND: Waiting for chain backend to finish sync, start_height=714884
2021-12-20 03:22:50.446 [INF] LNWL: Started rescan from block 000000000000000000035fba7bdfed50a131d12685d9edb0ba0ae2f95993a0cd (height 714555) for 0 addresses
2021-12-20 03:22:50.458 [CRT] CHDB: Caught unhandled error: context canceled
2021-12-20 03:22:50.458 [INF] CHDB: Sending request for shutdown
2021-12-20 03:22:50.458 [INF] LTND: Received shutdown request.
2021-12-20 03:22:50.458 [INF] LTND: Shutting down...
2021-12-20 03:22:50.458 [INF] LTND: Gracefully shutting down.
2021-12-20 03:22:50.458 [ERR] LNWL: Unable to process chain backend block connected notification: context canceled
2021-12-20 03:22:50.465 [INF] LNWL: Catching up block hashes to height 714562, this might take a while
2021-12-20 03:22:50.566 [INF] LNWL: Done catching up block hashes
2021-12-20 03:22:50.566 [INF] LNWL: Rescanned through block 000000000000000000081f53c492703c785f6fff23a92aa7325ff5603682e4f7 (height 714562)
2021-12-20 03:22:51.420 [INF] RPCS: Stopping RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping RouterRPC Sub-RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping InvoicesRPC Sub-RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping VersionRPC Sub-RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping WatchtowerRPC Sub-RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping WatchtowerClientRPC Sub-RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping SignRPC Sub-RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping WalletKitRPC Sub-RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping AutopilotRPC Sub-RPC Server
2021-12-20 03:22:51.420 [INF] RPCS: Stopping ChainRPC Sub-RPC Server
2021-12-20 03:22:51.421 [INF] LTND: Shutdown complete
```
Logs when using postgres v14.1
```
2021-12-20 04:11:56.336 [INF] LTND: We're not running within systemd
2021-12-20 04:11:56.336 [INF] LTND: Waiting for wallet encryption password. Use `lncli create` to create a wallet, `lncli unlock` to unlock an existing wallet, or `lncli changepassword` to change the password of an existing wallet and unlock it.
2021-12-20 04:12:28.182 [INF] LNWL: Opened wallet
2021-12-20 04:12:28.234 [INF] CHRE: Primary chain is set to: bitcoin
2021-12-20 04:12:28.246 [INF] CHRE: Initializing bitcoind backed fee estimator in CONSERVATIVE mode
2021-12-20 04:12:28.246 [INF] LNWL: Started listening for bitcoind transaction notifications via ZMQ on 100.0.10.82:28333
2021-12-20 04:12:28.246 [INF] LNWL: Started listening for bitcoind block notifications via ZMQ on 100.0.10.82:28332
2021-12-20 04:12:28.909 [INF] LNWL: The wallet has been unlocked without a time limit
2021-12-20 04:12:28.920 [CRT] CHDB: Caught unhandled error: context canceled
2021-12-20 04:12:28.921 [INF] CHDB: Sending request for shutdown
2021-12-20 04:12:28.921 [INF] LTND: Received shutdown request.
2021-12-20 04:12:28.921 [INF] LTND: Shutting down...
2021-12-20 04:12:28.921 [INF] LTND: Gracefully shutting down.
2021-12-20 04:12:28.921 [ERR] LTND: unable to create chain control: unable to create wallet: context canceled
2021-12-20 04:12:28.921 [ERR] LTND: Shutting down because error in main method: error loading chain control: unable to create chain control: unable to create wallet: context canceled
2021-12-20 04:12:28.922 [INF] LTND: Shutdown complete

error loading chain control: unable to create chain control: unable to create wallet: context canceled
```

### PR DESCRIPTION
**Pull request by joostjager**
fixes #6103

### CODE REVIEW

File: kvdb/postgres/readwrite_tx.go [hunk 1]
```diff
@@ -158,11 +158,11 @@ func (tx *readWriteTx) OnCommit(cb func()) {
 }
 
 // QueryRow executes a QueryRow call with a timeout context.
-func (tx *readWriteTx) QueryRow(query string, args ...interface{}) *sql.Row {
-	ctx, cancel := tx.db.getTimeoutCtx()
-	defer cancel()
+func (tx *readWriteTx) QueryRow(query string, args ...interface{}) (*sql.Row,
```
Comment (guggero) [inline kvdb/postgres/readwrite_tx.go:161]: Should we get rid of the `Query()` function below as it's currently not used but would present the same problem probably?

File: kvdb/postgres/readwrite_bucket.go [hunk 2]
```diff
@@ -187,19 +195,22 @@ func (b *readWriteBucket) CreateBucketIfNotExists(key []byte) (
...
 	switch {
 	// Bucket does not yet exist, so create it now. Postgres will generate a
 	// bucket id for the new bucket.
 	case err == sql.ErrNoRows:
-		err = b.tx.QueryRow(
+		row, cancel := b.tx.QueryRow(
 			"INSERT INTO "+b.table+" (parent_id, key) "+
-				"VALUES($1, $2) RETURNING id", b.id, key).
-			Scan(&id)
+				"VALUES($1, $2) RETURNING id", b.id, key)
```
Comment (guggero) [inline kvdb/postgres/readwrite_bucket.go:211]: nit: put closing parenthesis on next line.

**Review by guggero — CHANGES_REQUESTED** [review]
Changes LGTM, thanks for the fix!

Since this requires a bump in the `kvdb` module version, should we just merge this PR as a whole and close #6104? Then we don't need to wait for the rebase and don't accidentally push the tag before this PR is merged.

File: kvdb/postgres/readwrite_tx.go (hunk 1 above)
Comment (joostjager) [reply kvdb/postgres/readwrite_tx.go:161]: Good point, removed.

File: kvdb/postgres/readwrite_bucket.go (hunk 2 above)
Comment (joostjager) [reply kvdb/postgres/readwrite_bucket.go:211]: fixed

File: kvdb/postgres/readwrite_tx.go (hunk 1 above)
Comment (bhandras) [reply kvdb/postgres/readwrite_tx.go:161]: +1

**Review by bhandras — APPROVED** [review]
LGTM, thanks for the fix! :+1:

**Review by guggero — APPROVED** [review]

**Pull request merged**